src/tls.rs

```diff
@@ -395,12 +395,48 @@ impl Handshake {
         map_result_ssl(self, unsafe { SSL_do_handshake(self.as_ptr()) })
     }
 
-    pub fn get_write_level(&self) -> crypto::Level {
+    pub fn write_level(&self) -> crypto::Level {
         unsafe { SSL_quic_write_level(self.as_ptr()) }
     }
 
-    pub fn cipher(&self) -> Result<crypto::Algorithm> {
-        get_cipher_from_ptr(self.as_ptr())
+    pub fn cipher(&self) -> Option<crypto::Algorithm> {
+        get_cipher_from_ptr(self.as_ptr()).ok()
+    }
+
+    pub fn curve(&self) -> Option<String> {
+        let curve = unsafe {
+            let curve_id = SSL_get_curve_id(self.as_ptr());
+            if curve_id == 0 {
+                return None;
+            }
+
+            let curve_name = SSL_get_curve_name(curve_id);
+            match std::ffi::CStr::from_ptr(curve_name).to_str() {
+                Ok(v) => v,
+
+                Err(_) => return None,
+            }
+        };
+
+        Some(curve.to_string())
+    }
+
+    pub fn sigalg(&self) -> Option<String> {
+        let sigalg = unsafe {
+            let sigalg_id = SSL_get_peer_signature_algorithm(self.as_ptr());
+            if sigalg_id == 0 {
+                return None;
+            }
+
+            let sigalg_name = SSL_get_signature_algorithm_name(sigalg_id, 1);
+            match std::ffi::CStr::from_ptr(sigalg_name).to_str() {
```

Comment:
since we don't seem to care about the Ok case could we make this 
```
if std::ffi::CStr::from_ptr(sigalg_name).to_str().is_err() {
  return None;
}
```

Reply:
We do need the return value. The `match` "return" value in the `Ok(v)` case is then "returned" into `let curve = ...`.

Reply:
Oops!

Reply:
Oops!